sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -151,6 +164,30 @@ abstract class Expression extends TreeNode[Expression] {
   }
 
   /**
+   * Returns the hash for this expression. Expressions that compute the same result, even if
+   * they differ cosmetically should return the same hash.
+   */
+  def semanticHash() : Int = {
+    def computeHash(e: Seq[Any]): Int = {
```

Comment:
Whats wrong with `hashCode` on `Seq[Any]`?  I think this its just going to use [this](https://github.com/scala/scala/blob/27da46343cd545534819300235bc64ab74958c92/src/library/scala/util/hashing/MurmurHash3.scala#L217) which seems pretty reasonable.  Maybe I'm missing something though.